### PR TITLE
Fix duplicate Embind symbol

### LINKED ISSUE #20660: Duplicate symbol link error in 3.1.48 when including bind.h in more than one source file.  Ok in 3.1.46.
I am compiling my projects using the Docker images [here](https://hub.docker.com/r/emscripten/emsdk), as referenced in [the docs](https://emscripten.org/docs/getting_started/downloads.html#using-the-docker-image).  I'm running Docker Desktop on Windows 11(WSL2).

Below is the output of my build, first under 3.1.46, then under 3.1.48.

Following that are the sources I created to reproduce this, and have attached a zip file with all files, including a makefile and verbose build logs.  You can run 'make clean' to clear the cache, and 'make VERBOSE=true' to set EMCC_DEBUG := 2.  And 'make TYPE=opt' will do an optimized build.
[linkerr.zip](https://github.com/emscripten-core/emscripten/files/13301658/linkerr.zip)

I've included the makefile to show clearly the use all of the flags we're using in our project - I've tried removing all of these except -std=c++20 and -v, and 3.1.48 still gives the link error.

### Build under 3.1.46:
```
root@a8c172869196:/mnt/linkerr# emcc --version
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.46 (19607820c447a13fd8d0b7680c56148427d6e1b8)
Copyright (C) 2014 the Emscripten authors (see AUTHORS.txt)
This is free and open source software under the MIT license.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

root@a8c172869196:/mnt/linkerr# 
root@a8c172869196:/mnt/linkerr# make
em++  -std=c++20 -D__linux -Wno-backslash-newline-escape -Wno-bitwise-op-parentheses -Wno-deprecated-register -Wno-inconsistent-missing-override -Wno-logical-op-parentheses -Wmissing-declarations -Wdefaulted-function-deleted -Wno-unused-value -g -fwasm-exceptions   -c -o file1.o file1.cpp
em++  -std=c++20 -D__linux -Wno-backslash-newline-escape -Wno-bitwise-op-parentheses -Wno-deprecated-register -Wno-inconsistent-missing-override -Wno-logical-op-parentheses -Wmissing-declarations -Wdefaulted-function-deleted -Wno-unused-value -g -fwasm-exceptions   -c -o file2.o file2.cpp
  Linking 
 "/emsdk/upstream/bin/clang" --version
cache:INFO: generating system asset: symbol_lists/0193838d629247ad34612c7ac6c8b5797e3de0b2.json... (this will be cached in "/emsdk/upstream/emscripten/cache/symbol_lists/0193838d629247ad34612c7ac6c8b5797e3de0b2.json" for subsequent builds)
 "/emsdk/node/16.20.0_64bit/bin/node" /emsdk/upstream/emscripten/src/compiler.js /tmp/tmp3_jj52we.json --symbols-only
cache:INFO:  - ok
 "/emsdk/upstream/bin/wasm-ld" -o out.wasm ./file1.o ./file2.o -L/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten --whole-archive -lembind-rtti --no-whole-archive -lGL-webgl2-full_es3 -lal -lhtml5 -lstubs-debug -lnoexit -lc-debug -ldlmalloc-debug -lcompiler_rt-wasm-sjlj -lc++-except -lc++abi-debug-except -lunwind-except -lsockets -mllvm -combiner-global-alias-analysis=false -mllvm -wasm-enable-sjlj -mllvm -disable-lsr -mllvm -wasm-enable-eh -mllvm -exception-model=wasm /tmp/tmpp9lat0t3libemscripten_js_symbols.so --export-if-defined=main --export-if-defined=__get_exception_message --export-if-defined=free --export-if-defined=__cpp_exception --export-if-defined=__cxa_increment_exception_refcount --export-if-defined=__cxa_decrement_exception_refcount --export-if-defined=__thrown_object_from_unwind_exception --export-if-defined=__start_em_asm --export-if-defined=__stop_em_asm --export-if-defined=__start_em_lib_deps --export-if-defined=__stop_em_lib_deps --export-if-defined=__start_em_js --export-if-defined=__stop_em_js --export-if-defined=__main_argc_argv --export-if-defined=fflush --export=sbrk --export=emscripten_stack_get_end --export=emscripten_stack_get_free --export=emscripten_stack_get_base --export=emscripten_stack_get_current --export=emscripten_stack_init --export=stackSave --export=stackRestore --export=stackAlloc --export=__errno_location --export=emscripten_get_sbrk_ptr --export=__get_temp_ret --export=__set_temp_ret --export=__trap --export=__wasm_call_ctors --export=__get_exception_message --export=free --export=__thrown_object_from_unwind_exception --export=__cxa_increment_exception_refcount --export=__cxa_decrement_exception_refcount --export=malloc --export-table --growable-table -z stack-size=65536 --initial-memory=1073741824 --max-memory=4294967296 --no-entry --stack-first --table-base=1
 "/emsdk/upstream/bin/llvm-objcopy" out.wasm out.wasm --remove-section=producers
 "/emsdk/upstream/bin/wasm-emscripten-finalize" -g --dyncalls-i64 --pass-arg=legalize-js-interface-exported-helpers --dwarf out.wasm -o out.wasm --detect-features
 "/emsdk/node/16.20.0_64bit/bin/node" /emsdk/upstream/emscripten/src/compiler.js /tmp/tmp93zwfp28.json
emcc: warning: running limited binaryen optimizations because DWARF info requested (or indirectly required) [-Wlimited-postlink-optimizations]
 "/emsdk/upstream/bin/wasm-opt" --safe-heap out.wasm -o out.wasm -g --mvp-features --enable-exception-handling --enable-mutable-globals --enable-sign-ext
 "/emsdk/node/16.20.0_64bit/bin/node" /emsdk/upstream/emscripten/tools/acorn-optimizer.js /tmp/emscripten_temp__ohs40f2/out.js unsignPointers --closureFriendly --exportES6 -o /tmp/emscripten_temp__ohs40f2/out.jso1.js
 "/emsdk/node/16.20.0_64bit/bin/node" /emsdk/upstream/emscripten/tools/acorn-optimizer.js /tmp/emscripten_temp__ohs40f2/out.jso1.js safeHeap --closureFriendly --exportES6 -o /tmp/emscripten_temp__ohs40f2/out.jso2.js
root@a8c172869196:/mnt/linkerr# 
```

### Build under 3.1.48:
```
root@9b690697f719:/mnt/linkerr# emcc --version
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.48 (e967e20b4727956a30592165a3c1cde5c67fa0a8)
Copyright (C) 2014 the Emscripten authors (see AUTHORS.txt)
This is free and open source software under the MIT license.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

root@9b690697f719:/mnt/linkerr# make
em++  -std=c++20 -D__linux -Wno-backslash-newline-escape -Wno-bitwise-op-parentheses -Wno-deprecated-register -Wno-inconsistent-missing-override -Wno-logical-op-parentheses -Wmissing-declarations -Wdefaulted-function-deleted -Wno-unused-value -g -fwasm-exceptions   -c -o file1.o file1.cpp
em++  -std=c++20 -D__linux -Wno-backslash-newline-escape -Wno-bitwise-op-parentheses -Wno-deprecated-register -Wno-inconsistent-missing-override -Wno-logical-op-parentheses -Wmissing-declarations -Wdefaulted-function-deleted -Wno-unused-value -g -fwasm-exceptions   -c -o file2.o file2.cpp
  Linking 
 /emsdk/upstream/bin/clang --version
cache:INFO: generating system asset: symbol_lists/c80c1b80deac4f1d31b53986d27fdd418865c424.json... (this will be cached in "/emsdk/upstream/emscripten/cache/symbol_lists/c80c1b80deac4f1d31b53986d27fdd418865c424.json" for subsequent builds)
 /emsdk/node/16.20.0_64bit/bin/node /emsdk/upstream/emscripten/src/compiler.js /tmp/tmpizzavjf7.json --symbols-only
cache:INFO:  - ok
 /emsdk/upstream/bin/wasm-ld -o out.wasm ./file1.o ./file2.o -L/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten --whole-archive -lembind-rtti --no-whole-archive -lGL-webgl2-full_es3 -lal -lhtml5 -lstubs-debug -lnoexit -lc-debug -ldlmalloc-debug -lcompiler_rt-wasm-sjlj -lc++-except -lc++abi-debug-except -lunwind-except -lsockets -mllvm -combiner-global-alias-analysis=false -mllvm -wasm-enable-sjlj -mllvm -disable-lsr -mllvm -wasm-enable-eh -mllvm -exception-model=wasm /tmp/tmppk7ewi1ilibemscripten_js_symbols.so --export-if-defined=main --export-if-defined=__get_exception_message --export-if-defined=free --export-if-defined=__cpp_exception --export-if-defined=__cxa_increment_exception_refcount --export-if-defined=__cxa_decrement_exception_refcount --export-if-defined=__thrown_object_from_unwind_exception --export-if-defined=__start_em_asm --export-if-defined=__stop_em_asm --export-if-defined=__start_em_lib_deps --export-if-defined=__stop_em_lib_deps --export-if-defined=__start_em_js --export-if-defined=__stop_em_js --export-if-defined=__main_argc_argv --export-if-defined=fflush --export=sbrk --export=emscripten_stack_get_end --export=emscripten_stack_get_free --export=emscripten_stack_get_base --export=emscripten_stack_get_current --export=emscripten_stack_init --export=stackSave --export=stackRestore --export=stackAlloc --export=__errno_location --export=emscripten_get_sbrk_ptr --export=__get_temp_ret --export=__set_temp_ret --export=__trap --export=__wasm_call_ctors --export=__get_exception_message --export=free --export=__thrown_object_from_unwind_exception --export=__cxa_increment_exception_refcount --export=__cxa_decrement_exception_refcount --export=malloc --export-table --growable-table -z stack-size=65536 --initial-memory=1073741824 --max-memory=4294967296 --no-entry --stack-first --table-base=1
wasm-ld: error: duplicate symbol: _emval_coro_resume
>>> defined in ./file1.o
>>> defined in ./file2.o
emcc: error: '/emsdk/upstream/bin/wasm-ld -o out.wasm ./file1.o ./file2.o -L/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten --whole-archive -lembind-rtti --no-whole-archive -lGL-webgl2-full_es3 -lal -lhtml5 -lstubs-debug -lnoexit -lc-debug -ldlmalloc-debug -lcompiler_rt-wasm-sjlj -lc++-except -lc++abi-debug-except -lunwind-except -lsockets -mllvm -combiner-global-alias-analysis=false -mllvm -wasm-enable-sjlj -mllvm -disable-lsr -mllvm -wasm-enable-eh -mllvm -exception-model=wasm /tmp/tmppk7ewi1ilibemscripten_js_symbols.so --export-if-defined=main --export-if-defined=__get_exception_message --export-if-defined=free --export-if-defined=__cpp_exception --export-if-defined=__cxa_increment_exception_refcount --export-if-defined=__cxa_decrement_exception_refcount --export-if-defined=__thrown_object_from_unwind_exception --export-if-defined=__start_em_asm --export-if-defined=__stop_em_asm --export-if-defined=__start_em_lib_deps --export-if-defined=__stop_em_lib_deps --export-if-defined=__start_em_js --export-if-defined=__stop_em_js --export-if-defined=__main_argc_argv --export-if-defined=fflush --export=sbrk --export=emscripten_stack_get_end --export=emscripten_stack_get_free --export=emscripten_stack_get_base --export=emscripten_stack_get_current --export=emscripten_stack_init --export=stackSave --export=stackRestore --export=stackAlloc --export=__errno_location --export=emscripten_get_sbrk_ptr --export=__get_temp_ret --export=__set_temp_ret --export=__trap --export=__wasm_call_ctors --export=__get_exception_message --export=free --export=__thrown_object_from_unwind_exception --export=__cxa_increment_exception_refcount --export=__cxa_decrement_exception_refcount --export=malloc --export-table --growable-table -z stack-size=65536 --initial-memory=1073741824 --max-memory=4294967296 --no-entry --stack-first --table-base=1' failed (returned 1)
make: *** [makefile:76: all] Error 1
```

### file1.cpp
**Just remove the bind.h include and it will also link under 3.1.48.**
```cpp
#include <emscripten/emscripten.h>
#include <emscripten/bind.h>

EM_JS(void, file1testJS, (), {
    console.log ("file1testJS");
});

void file1test() {
    file1testJS();
}

extern void file2test();

int main() {
    file1test();
    file2test();
    return 0;
}
```

### file2.cpp
```cpp
#include <emscripten/emscripten.h>
#include <emscripten/bind.h>

EM_JS(void, file2testJS, (), {
    console.log ("file2testJS");
});

void file2test() {
    file2testJS();
}
```

### PR DESCRIPTION
**Pull request by RReverser**
Fixes #20660.

### CODE REVIEW

**Review by sbc100 — APPROVED** [review]

**Comment by RReverser** [comment]
Okay a quick fix didn't work, apparently `inline` makes it omitted from linking altogether...

I can't move it to bind.cpp either because it depends on C++ version the val.h header is used with.

The only remaining option that I can think of is to mark it as weakly referenced. @sbc100 do you have better ideas?

**Comment by RReverser** [comment]
OTOH I guess deduplicating symbols is exactly what weak linking is for, so maybe not that dirty after all.

File: system/include/emscripten/val.h [hunk 1]
```diff
@@ -708,6 +708,7 @@ class val_awaiter {
 
 extern "C" {
   // JS FFI helper for `val_awaiter::resume_with`.
+  __attribute__((weak))
   void _emval_coro_resume(val_awaiter* awaiter, EM_VAL result) {
```
Comment (sbc100) [inline system/include/emscripten/val.h:712]: I believe `static inline` is the more common way to do this.

Comment (sbc100) [reply system/include/emscripten/val.h:712]: Or wait you need to export it.. ignore me,

**Comment by sbc100** [comment]
Can we just move this function into the `system/lib/embind/bind.cpp` and just make sure that library is built with C++20?

File: system/include/emscripten/val.h (hunk 1 above)
Comment (RReverser) [reply system/include/emscripten/val.h:712]: Ah I didn't realise it's diferent from just `inline`. I need to read up and refresh on differences...

**Comment by RReverser** [comment]
> Can we just move this function into the `system/lib/embind/bind.cpp` and just make sure that library is built with C++20?

Hm... I guess we could, but I'm not sure if linker will manage to eliminate all the co_await machinery if the operator is actually not used.

File: system/include/emscripten/val.h (hunk 1 above)
Comment (RReverser) [reply system/include/emscripten/val.h:712]: Nope, it doesn't work - just like with regular `inline`, it results in symbol not found.

**Comment by sbc100** [comment]
> > Can we just move this function into the `system/lib/embind/bind.cpp` and just make sure that library is built with C++20?
> 
> Hm... I guess we could, but I'm not sure if linker will manage to eliminate all the co_await machinery if the operator is actually not used.

This symbol is pulled in by a JS dependency.. the linker will only include it if the JS symbol is used.   Putting it the C++ wont' make that any better or worse IIUC.

**Comment by RReverser** [comment]
> This symbol is pulled in by a JS dependency.. the linker will only include it if the JS symbol is used.

Right, but JS dependency is in turn pulled by C++, so there's sort of a cycle there. That said, it all ends up in `val_awaiter::await_suspend`, so I guess as long as C++ doesn't implicitly link it in and uses those "well-known" methods only when `co_await` is actually used, I guess it should be fine.

**Comment by RReverser** [comment]
I think I'll merge the simple working solution first, and look into improving it separately.

**Comment by sbc100** [comment]
> > This symbol is pulled in by a JS dependency.. the linker will only include it if the JS symbol is used.
> 
> Right, but JS dependency is in turn pulled by C++, so there's sort of a cycle there. That said, it all ends up in `val_awaiter::await_suspend`, so I guess as long as C++ doesn't implicitly link it in and uses those "well-known" methods only when `co_await` is actually used, I guess it should be fine.

I mean the linker either requires the symbol is it doesn't.  Whether its defined in a header or not makes no difference to the linker.  It can't tell the difference.

**Comment by sbc100** [comment]
lgtm either way.  

Anther reason to move it the cpp file is to avoid paying the compilation cost for ever source that includes this header.

**Comment by RReverser** [comment]
> Whether its defined in a header or not makes no difference to the linker. It can't tell the difference.

Right, but what I meant is that there can be potentially a difference if we build bind.cpp with C++20 because then it always includes those extra symbols from val.h, whereas right now they're only included if an actual consumer builds with C++20.

So in one case those symbols don't even appear in the compilation pipeline, while in another they're always there as part of bind.o and it's a job of the linker to exclude them.

**Comment by RReverser** [comment]
> Anther reason to move it the cpp file is to avoid paying the compilation cost for ever source that includes this header.

That's true, but given how tiny that function is, it doesn't seem any different from all other functions already in val.h. You said yourself recently that those such small functions should probably stay in val.h so that they're always inlined.

**Pull request merged**